Use smart pointers to store reference to CanvasBase in CanvasRenderingContext

iidmsa opened the pull request:
#### ddfb6b7d179b45962c4a7f15d264417f9601a30f
<pre>
Use smart pointers to store reference to CanvasBase in CanvasRenderingContext
<a href="https://bugs.webkit.org/show_bug.cgi?id=286535">https://bugs.webkit.org/show_bug.cgi?id=286535</a>
<a href="https://rdar.apple.com/problem/143630378">rdar://problem/143630378</a>

Reviewed by Chris Dumez and Darin Adler.

Fix a clang static analyzer warning caused by using a raw reference to
store the owning CanvasBase in CanvasRenderingContext. Since CanvasBase
owns CanvasRenderingContext I resorted to using a `WeakRef`.

* Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations:
* Source/WebCore/html/CanvasBase.h:
* Source/WebCore/html/HTMLCanvasElement.h:
* Source/WebCore/html/canvas/CanvasRenderingContext.cpp:
(WebCore::CanvasRenderingContext::ref const):
(WebCore::CanvasRenderingContext::deref const):
(WebCore::CanvasRenderingContext::taintsOrigin):
(WebCore::CanvasRenderingContext::checkOrigin):
* Source/WebCore/html/canvas/CanvasRenderingContext.h:
(WebCore::CanvasRenderingContext::checkOrigin):

Canonical link: <a href="https://commits.webkit.org/289414@main">https://commits.webkit.org/289414@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a7b5db42b755e9fb445fd79c207d4a5891b884bb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/86877 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6384 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41219 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/91725 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/37609 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/88926 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/6652 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/14451 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67157 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24932 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/89880 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5066 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78633 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47476 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4851 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33003 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/36727 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/75356 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/33875 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/93618 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14030 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10191 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/75960 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14231 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74478 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75157 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18477 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19475 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/17890 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/6864 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14053 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/19315 "Failed to build and analyze WebKit") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/13791 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17236 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/15576 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/html/canvas/CanvasRenderingContext.h:
```diff
@@ -157,7 +157,7 @@ class CanvasRenderingContext : public ScriptWrappable, public CanMakeWeakPtr<Can
     static Lock s_instancesLock;
 
-    CanvasBase& m_canvas;
+    Ref<CanvasBase> m_canvas;
```
This is wrong. The CanvasBase owns the CanvasRenderingContext. Now it will ref its owner, resulting in a reference cycle in thus a memory leak. We likely want to use `WeakRef<>` instead, which will require some changes to `CanvasBase`. Replacing a `&` with a `Ref<>` is rarely a good idea.

iidmsa replied:
Changes applied as instructed. Thank you very much for your comment and time.

darinadler commented on Source/WebCore/html/CanvasBase.h:
```diff
@@ -72,14 +73,14 @@ class CanvasDisplayBufferObserver : public CanMakeWeakPtr<CanvasDisplayBufferObs
-    void ref() const { refCanvasBase(); }
-    void deref() const { derefCanvasBase(); }
+    void ref() const override { refCanvasBase(); }
+    void deref() const override { derefCanvasBase(); }
```
```suggestion
    void ref() const final { refCanvasBase(); }
    void deref() const final { derefCanvasBase(); }
```

darinadler commented on Source/WebCore/html/CanvasBase.h:
```diff
@@ -72,14 +73,14 @@ class CanvasDisplayBufferObserver : public CanMakeWeakPtr<CanvasDisplayBufferObs
     virtual ~CanvasBase();
 
     virtual void refCanvasBase() const = 0;
     virtual void derefCanvasBase() const = 0;
```
It's possible we can also get rid of these and just use `ref`/`deref` as the virtual functions since we are adding virtual versions of them. Might be a nice cleanup and would be slightly better for performance.

cdumez reviewed (approved):
LGTM

iidmsa committed:
Use smart pointers to store reference to CanvasBase in CanvasRenderingContext
https://bugs.webkit.org/show_bug.cgi?id=286535
rdar://problem/143630378

Reviewed by Chris Dumez and Darin Adler.

Fix a clang static analyzer warning caused by using a raw reference to
store the owning CanvasBase in CanvasRenderingContext. Since CanvasBase
owns CanvasRenderingContext I resorted to using a `WeakRef`.

* Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations:
* Source/WebCore/html/CanvasBase.h:
* Source/WebCore/html/HTMLCanvasElement.h:
* Source/WebCore/html/canvas/CanvasRenderingContext.cpp:
(WebCore::CanvasRenderingContext::ref const):
(WebCore::CanvasRenderingContext::deref const):
(WebCore::CanvasRenderingContext::taintsOrigin):
(WebCore::CanvasRenderingContext::checkOrigin):
* Source/WebCore/html/canvas/CanvasRenderingContext.h:
(WebCore::CanvasRenderingContext::checkOrigin):

Canonical link: https://commits.webkit.org/289414@main
